src/index.js

```diff
@@ -0,0 +1,159 @@
+
+let translate = NotInitialisedError;
+let formatDate = NotInitialisedError;
+let Translation = NotInitialisedError;
+
+const withI18n = Component => {
+  class WithI18nComponent extends PureComponent {
+    render() {
+      return <Component {...this.props} translate={translate} formatDate={formatDate} />;
+    }
+  }
+
+  return WithI18nComponent;
+};
+
+class Provider extends PureComponent {
+  state = {
+    loaded: false,
+  };
+
+  async componentDidMount() {
+    const { domain } = this.props;
+    const language = this.getUserLanguage();
+    const localeData = this.getAllLocaleData();
+
+    addLocaleData(localeData);
+    const translations = await this.fetchTranslations(language);
+
+    const component = React.createElement(IntlProvider, {
+      messages: translations,
+      children: React.createElement('div'),
+      locale: language,
+      ref: element => {
+        const { intl } = element.getChildContext();
+
+        const provideIntlContext = Component => {
+          class IntlComponent extends PureComponent {
+            static childContextTypes = {
+              intl: PropTypes.object,
+            };
+
+            static propTypes = {
+              id: PropTypes.string.isRequired,
+            };
+
+            getChildContext() {
+              return { intl };
+            }
+
+            render() {
+              const { id, ...others } = this.props;
+              return <Component {...others} id={`${domain}.${id}`} />;
+            }
+          }
+
+          return IntlComponent;
+        };
+
+        translate = (id, values) => intl.formatMessage({ id: domain ? `${domain}.${id}` : id }, values);
+        formatDate = intl.formatDate;
+        Translation = provideIntlContext(FormattedMessage);
+
+        this.setState({ loaded: true });
+      },
+    });
+
+    render(component, document.createElement('div'));
+  }
+
+  getUserLanguage() {
+    if (this.props.isDevelopmentMode) {
+      return this.props.developmentLanguage;
+    }
+
+    const language = document.body.parentNode.getAttribute('lang');
+
+    if (!supportedLanguages.includes(language)) {
+      return this.props.fallbackLanguage;
+    }
+
+    return language;
+  }
+
+  async fetchTranslations(language) {
+    const path = this.getTranslationsPath(language);
+
+    try {
+      const response = await fetch(path);
+      const json = await response.json();
+      return json;
+    } catch (error) {}
+
+    return {};
+  }
+
+  getTranslationsPath(language) {
+    const { path } = this.props;
+
+    if (typeof path === 'function') {
+      return path(language);
+    }
+
+    return `${path}${language}.json`;
+  }
+
+  getAllLocaleData() {
+    return supportedLanguages.reduce((currentLocaleData, language) => {
+      const languageCode = language.split('-')[0];
+      const localeData = this.getLocaleDataForLanguage(languageCode);
+      return [...currentLocaleData, ...localeData];
+    }, []);
```

Comment:
PP: I think you can use `Array.prototype.map` here instead of `reduce`:
```
return supportedLanguages.map(language => {
    const languageCode = language.split('-')[0];
    return this.getLocaleDataForLanguage(languageCode);
});
```

Reply:
That will give me an array of arrays instead of 1 big array, right?

Reply:
Ah yes, I looked over the fact that `localeData` was also spread into the result. Indeed map wouldn't do that.